File: lintcode/python/0547_intersection_of_two_arrays.py
#!/usr/bin/python3 -t

# sort and two pointers
# time O(max(nlogn, mlogm))
# space O(1)

class Solution:
    """
    @param nums1: an integer array
    @param nums2: an integer array
    @return: an integer array
    """
    def intersection(self, nums1, nums2):
        # write your code here
        m = len(nums1)
        n = len(nums2)
        if n == 0 or m == 0:
            return []

        nums1.sort()
        nums2.sort()

        ret = set()
        i, j = 0, 0

        while i < m and j < n:
            if nums1[i] == nums2[j]:
                ret.add(nums1[i])
                i += 1
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                j  += 1

        return list(ret)


if __name__ == '__main__':
    s = Solution()
    a = [1,2,2,1]
    b = [2,2]
    print(s.intersection(a, b))

# sort and binary search solution

class Solution:
    """
    @param nums1: an integer array
    @param nums2: an integer array
    @return: an integer array
    """
    def intersection(self, nums1, nums2):
        # write your code here
        m = len(nums1)
        n = len(nums2)
        if m == 0 or n == 0:
            return []
        nums1.sort()
        nums2.sort()
        
        ret = set()
        
        if m < n:
            nums1, nums2 = nums2, nums1
            m, n = n, m
            
        for i in range(n):
            if self.found(nums1, nums2[i]):
                ret.add(nums2[i])
                
        return list(ret)
        
    def found(self, nums, target):
        n = len(nums)
        l = 0
        r = n-1
        while l<r:
            mid = (l+r)/2
            if nums[mid] == target:
                return True
            elif nums[mid] < target:
                l = mid +1
            else:
                r = mid
                
        if nums[l] == target:
            return True
        return False

File: lintcode/golang/0547_intersection_of_two_arrays.go
/**
 * @param nums1: an integer array
 * @param nums2: an integer array
 * @return: an integer array
 */
func intersection (nums1 []int, nums2 []int) []int {
    // write your code here
    h := make(map[int]bool)
    ret := make([]int, 0)
    
    for _, val := range nums1 {
        h[val] = true
    }
    
    for _, val := range nums2 {
        if h[val] {
            ret = append(ret, val)
            h[val] = false
        }
    }
    
    quickSort(&ret, 0, len(ret)-1)
    
    return ret
}

func quickSort(ret *[]int, start int, end int) {
    if start < end {
        part := getPart(ret, start, end)
        
        quickSort(ret, start, part-1)
        quickSort(ret, part+1, end)
    }
}

func getPart(ret *[]int, start int, end int) int {
    pivot := (*ret)[end]
    index := start
    
    for i:= start; i < end; i++ {
        if (*ret)[i] <= pivot {
            tmp := (*ret)[i]
            (*ret)[i] = (*ret)[index]
            (*ret)[index] = tmp
            index += 1
        }
    }
    
    tmp := (*ret)[end]
    (*ret)[end] = (*ret)[index]
    (*ret)[index] = tmp
    return index
}
